File: ios/flutter_meta_wearables_dat/Sources/flutter_meta_wearables_dat/NativeVideoFrameConsumer.swift
import AVFoundation
import Foundation

/// Receives raw DAT frames without copying their pixels through Flutter.
///
/// Consumers must not retain the sample buffer after this callback returns.
public protocol MetaWearablesDatNativeVideoFrameConsumer: AnyObject {
  func metaWearablesDatDidOutput(sampleBuffer: CMSampleBuffer)
}

/// Process-local registry used by optional sibling Flutter plugins.
public final class MetaWearablesDatNativeVideoFrameConsumers: NSObject {
  private static let lock = NSLock()
  private static let consumers = NSHashTable<AnyObject>.weakObjects()

  public static func add(_ consumer: MetaWearablesDatNativeVideoFrameConsumer) {
    lock.lock()
    consumers.add(consumer)
    lock.unlock()
  }

  public static func remove(_ consumer: MetaWearablesDatNativeVideoFrameConsumer) {
    lock.lock()
    consumers.remove(consumer)
    lock.unlock()
  }

  static func dispatch(_ sampleBuffer: CMSampleBuffer) {
    lock.lock()
    let snapshot = consumers.allObjects
    lock.unlock()
    snapshot.forEach {
      ($0 as? MetaWearablesDatNativeVideoFrameConsumer)?
        .metaWearablesDatDidOutput(sampleBuffer: sampleBuffer)
    }
  }
}

File: android/src/main/kotlin/io/rodcone/flutter_meta_wearables_dat/NativeVideoFrameConsumer.kt
package io.rodcone.flutter_meta_wearables_dat

import java.nio.ByteBuffer
import java.util.concurrent.CopyOnWriteArraySet

/**
 * A synchronous, native-only view of a DAT video frame.
 *
 * Consumers must finish reading [buffer] before [onVideoFrame] returns. The
 * buffer is a read-only duplicate whose position is independent from DAT's
 * source buffer, but its backing memory is owned by the DAT SDK and may be
 * reused after the callback.
 */
public data class NativeVideoFrame(
        val buffer: ByteBuffer,
        val width: Int,
        val height: Int,
        val presentationTimeUs: Long,
        val isCompressed: Boolean,
)

/** Receives frames without routing their bytes through a Flutter event channel. */
public fun interface NativeVideoFrameConsumer {
    public fun onVideoFrame(frame: NativeVideoFrame)
}

/**
 * Process-local registry for optional native encoders, recorders, and vision
 * processors hosted by sibling Flutter plugins.
 */
public object NativeVideoFrameConsumers {
    private val consumers = CopyOnWriteArraySet<NativeVideoFrameConsumer>()

    @JvmStatic
    public fun add(consumer: NativeVideoFrameConsumer) {
        consumers.add(consumer)
    }

    @JvmStatic
    public fun remove(consumer: NativeVideoFrameConsumer) {
        consumers.remove(consumer)
    }

    internal fun dispatch(frame: com.meta.wearable.dat.camera.types.VideoFrame) {
        if (consumers.isEmpty()) return
        val nativeFrame =
                NativeVideoFrame(
                        buffer = frame.buffer.asReadOnlyBuffer(),
                        width = frame.width,
                        height = frame.height,
                        presentationTimeUs = frame.presentationTimeUs,
                        isCompressed = frame.isCompressed,
                )
        consumers.forEach { consumer -> consumer.onVideoFrame(nativeFrame) }
    }
}
